### PR TITLE
chore/otel-collector: upgrade to v0.103.0, remove jaegerexporter

### PR DESCRIPTION
**Pull request by bobheadxi**
Thread: https://sourcegraph.slack.com/archives/C1JH2BEHZ/p1717797870638299

One problem caused by this upgrade is that the [deprecated `jaegerexporter`](https://github.com/open-telemetry/opentelemetry-specification/pull/2858) no longer builds at all with the last published version, so for the upgrade to go through it must be removed. I've updated local `sg start` to work with this change, but some Release team support is needed for deployment configuration + customer-facing docs changes:  https://sourcegraph.slack.com/archives/C1JH2BEHZ/p1718143249191349?thread_ts=1717797870.638299&cid=C1JH2BEHZ, since current guidance asks customers to configure `jaegerexporter`.

Part of https://linear.app/sourcegraph/issue/SEC-1680

Closes https://linear.app/sourcegraph/issue/CORE-177

## Test plan

Followed steps shared in https://sourcegraph.slack.com/archives/C04MYFW01NV/p1718136211292469 to run locally, since `sg run jaeger otel-collector` alone is insufficient to get updated images:

```diff
diff --git a/wolfi-images/opentelemetry-collector.yaml b/wolfi-images/opentelemetry-collector.yaml
index b995a3d531b..76de80d4f84 100644
--- a/wolfi-images/opentelemetry-collector.yaml
+++ b/wolfi-images/opentelemetry-collector.yaml
@@ -7,7 +7,11 @@ contents:
     - mailcap
 
     ## opentelemetry-collector packages
-    - opentelemetry-collector@sourcegraph
+    - opentelemetry-collector@branch
+  keyring:
+    - https://packages.sgdev.org/sourcegraph-melange-dev.rsa.pub
+  repositories:
+    - '@branch https://packages.sgdev.org/branches/otel-collector-upgrade'
 
 paths:
   - path: /otel-collector
```

plus `sg wolfi lock opentelemetry-collector` will make `sg run otel-collector` use the correct image.

**The above diffs SHOULD NOT be committed. The lock should happen post-merge.**

Spot-checked some traces locally with:

```
sg run jaeger otel-collector
sg start
```

image

image

## Changelog 

- OpenTelemetry Collector: Upgraded OpenTelemetry Collector to v0.103.0
- OpenTelemetry Collector: The [deprecated `jaegerexporter` has been removed](https://github.com/open-telemetry/opentelemetry-specification/pull/2858). Users of `exporter: { jaeger: ... }` in the OpenTelemetry Collector should use `exporter: { otlp/jaeger: ... }` to send traces directly to Jaeger via its OTLP receiver.
   - Users of the default Jaeger configurations now need to provide `JAEGER_OTLP_GRPC_PORT` as well as the existing `JAEGER_HOST` configuration.

### CODE REVIEW

**Review by evict — APPROVED** [review]

**Review by unknwon — APPROVED** [review]

**Review by jhchabran — DISMISSED** [review]
This PR updates GRPC from 1.61 to 1.64 which deprecates the `.Dial*` functions. This is caught by the CI, and we most likely want to update it. 

Looking at https://github.com/jaegertracing/jaeger/issues/5330 it seems that there are a few little semantics differences to take in account, making this potentially not trivial. 

I'll ping the source team as @ggilmore has been leading this for a while.

File: doc/dependency_decisions.yml [hunk 1]
```diff
@@ -598,20 +598,27 @@
...
+  - :who: 
+    :why: 
     :versions: []
     :when: 2024-06-05 15:32:44.440111000 Z
+- - :license
+  - github.com/hashicorp/go-version
+  - MPL-2.0
+  - :who: 
+    :why: 
+    :versions: []
```
Comment (eseliger) [inline doc/dependency_decisions.yml:623]: this seems to be true for all things in here, and is unrelated to the PR, but I am confused why versions is empty :D

File: go.mod [hunk 2]
```diff
@@ -40,8 +40,6 @@ replace (
 	github.com/gomodule/redigo => github.com/gomodule/redigo v1.8.9
 	// Pending: Renamed to github.com/google/gnostic. Transitive deps still use the old name (kubernetes/kubernetes).
 	github.com/googleapis/gnostic => github.com/googleapis/gnostic v0.5.5
-	// Pending a release cut of https://github.com/prometheus/alertmanager/pull/3010
```
Comment (eseliger) [inline go.mod:43]: yes!!

File: go.mod [hunk 3]
```diff
@@ -56,15 +54,15 @@ replace (
...
 	cloud.google.com/go/profiler v0.4.0
-	cloud.google.com/go/pubsub v1.37.0
-	cloud.google.com/go/secretmanager v1.11.5
-	cloud.google.com/go/storage v1.38.0
-	github.com/GoogleCloudPlatform/opentelemetry-operations-go/exporter/metric v0.45.0
-	github.com/GoogleCloudPlatform/opentelemetry-operations-go/exporter/trace v1.21.0
+	cloud.google.com/go/pubsub v1.38.0
+	cloud.google.com/go/secretmanager v1.13.1
+	cloud.google.com/go/storage v1.40.0
+	github.com/GoogleCloudPlatform/opentelemetry-operations-go/exporter/metric v0.48.0
```
Comment (eseliger) [inline go.mod:64]: Sad that these all had to upgrade for this otel bump, makes it harder to track down regressions :)

**Review by eseliger — APPROVED** [review]

File: doc/dependency_decisions.yml (hunk 1 above)
Comment (bobheadxi) [reply doc/dependency_decisions.yml:623]: I think it's optional fields you can record when tracking a dependency change, e.g. "this code was MPL for vX to vY"

File: go.mod (hunk 3 above)
Comment (bobheadxi) [reply go.mod:64]: yeah the dependency tree is pretty wild. I did upgrade aggressively but we've ran into more issues with not upgrading in lockstep, than we have with upgrading everything OTEL depends on in one go

**Comment by bobheadxi** [comment]
I had to remove the Jaeger integration from our OTEL-collector build, because it has been deprecated for a while and now no longer builds at all - see changelog in my description, and thread: https://sourcegraph.slack.com/archives/C1JH2BEHZ/p1718142210823969?thread_ts=1717797870.638299&cid=C1JH2BEHZ

@sourcegraph/release might need your support here to land this. See https://sourcegraph.slack.com/archives/C1JH2BEHZ/p1717797870638299 for context on this upgrade

**Comment by jhchabran** [comment]
@bobheadxi just checking, as we're expected to review here. Still waiting on the release side?

**Comment by bobheadxi** [comment]
@jhchabran this is pending release team having some time to work with us on https://linear.app/sourcegraph/issue/REL-166 , and the immediate risks to dotcom have been mitigated by Security. I've split out some of the upgrades in this PR to other PRs, and now this is pretty out of date - I'll circle back to this when we're ready. Reverting to draft for now.

**Comment by Chickensoupwithrice** [comment]
I've been trying to test this with ephemeral deployments (see [here](https://sourcegraph.slack.com/archives/C06JENN2QBF/p1718699370323919)) but with failing tests, it won't get deployed. 

Though it looks like they're primarily time out related failures, I've just started them for a third time to confirm whether they're flakes or not. Being unfamiliar with this section of code, and the specific test failures makes it a little tricky for me to ensure smooth landing of this branch. But I'm looking into resolving these test failures. 

I'm also willing to step up and keep this branch up to date with main until the release team can complete work on https://linear.app/sourcegraph/issue/REL-166/cve-2024-36129-update-jaeger-and-otel-in-deploy-repos

**Comment by bobheadxi** [comment]
@Chickensoupwithrice thanks for taking this for a spin and offering to maintain it :) I'm a bit busy this week, but let's schedule a time next week to walk through keeping this up to date and how things work?

**Comment by Chickensoupwithrice** [comment]
@bobheadxi That sounds great! Thank you :)

File: sg.config.yaml [hunk 4]
```diff
@@ -821,12 +821,18 @@ commands:
...
+        --collector.otlp.grpc.host-port=:${JAEGER_OTLP_GRPC_PORT} \
+        --collector.otlp.http.host-port=:${JAEGER_OTLP_HTTP_PORT}
     install_func: installJaeger
     env:
       JAEGER_VERSION: 1.45.0
       JAEGER_DISK: $HOME/.sourcegraph-dev/data/jaeger
       JAEGER_LOG_LEVEL: error
+      COLLECTOR_OTLP_ENABLED: 'true'
+      JAEGER_OTLP_GRPC_PORT: 4320
+      JAEGER_OTLP_HTTP_PORT: 4321
```
Comment (bobheadxi) [inline sg.config.yaml:835]: We need to enable this on deploy-*

**Review by Chickensoupwithrice — APPROVED** [review]

File: internal/search/job/jobutil/exhaustive_job_test.go [hunk 5]
```diff
@@ -47,15 +47,15 @@ func TestNewExhaustive(t *testing.T) {
       (useFullDeadline . true)
       (patternInfo . TextPatternInfo{"content",case,nopath,filematchlimit:1000000,lang:cpp,F:"(?i)(?:\\.cpp$)|(?:\\.c\\+\\+$)|(?:\\.cc$)|(?:\\.cp$)|(?:\\.cppm$)|(?:\\.cxx$)|(?:\\.h$)|(?:\\.h\\+\\+$)|(?:\\.hh$)|(?:\\.hpp$)|(?:\\.hxx$)|(?:\\.inc$)|(?:\\.inl$)|(?:\\.ino$)|(?:\\.ipp$)|(?:\\.ixx$)|(?:\\.re$)|(?:\\.tcc$)|(?:\\.tpp$)|(?:\\.txx$)"})
       (numRepos . 0)
-      (pathRegexps . [(?i)(?:\.cpp$)|(?:\.c\+\+$)|(?:\.cc$)|(?:\.cp$)|(?:\.cppm$)|(?:\.cxx$)|(?:\.h$)|(?:\.h\+\+$)|(?:\.hh$)|(?:\.hpp$)|(?:\.hxx$)|(?:\.inc$)|(?:\.inl$)|(?:\.ino$)|(?:\.ipp$)|(?:\.ixx$)|(?:\.re$)|(?:\.tcc$)|(?:\.tpp$)|(?:\.txx$)])
+      (pathRegexps . ["(?i)(?:\\.cpp$)|(?:\\.c\\+\\+$)|(?:\\.cc$)|(?:\\.cp$)|(?:\\.cppm$)|(?:\\.cxx$)|(?:\\.h$)|(?:\\.h\\+\\+$)|(?:\\.hh$)|(?:\\.hpp$)|(?:\\.hxx$)|(?:\\.inc$)|(?:\\.inl$)|(?:\\.ino$)|(?:\\.ipp$)|(?:\\.ixx$)|(?:\\.re$)|(?:\\.tcc$)|(?:\\.tpp$)|(?:\\.txx$)"])
```
Comment (jhchabran) [inline internal/search/job/jobutil/exhaustive_job_test.go:50]: Why are we seeing those changes here? It seems to have nothing to do with the PR

Comment (jhchabran) [inline internal/search/job/jobutil/exhaustive_job_test.go:50]: Do we know where this is coming from? cc @varungandhi-src perhaps?

Comment (varungandhi-src) [reply internal/search/job/jobutil/exhaustive_job_test.go:50]: @jhchabran sorry, I don't know about this. I've flagged this in the search platform team channel.

Comment (stefanhengl) [reply internal/search/job/jobutil/exhaustive_job_test.go:50]: Apparently we use `go.opentelemetry.io/otel` to encode values in the s-exp printer for the tests (see [here](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/search/job/printer/sexp.go?L52)). 

Without looking at  the source of `go.opentelemetry.io/otel`, My best bet is they changed the way they process arrays.

Comment (stefanhengl) [reply internal/search/job/jobutil/exhaustive_job_test.go:50]: This looks related

https://sourcegraph.com/github.com/open-telemetry/opentelemetry-go/-/commit/f8b9fe3dbe676af208caa963f38a1a1e1f35c7be?visible=4

Comment (jhchabran) [reply internal/search/job/jobutil/exhaustive_job_test.go:50]: Thanks @stefanhengl. Do you think this is creating a problem here?

Comment (stefanhengl) [reply internal/search/job/jobutil/exhaustive_job_test.go:50]: I don't think so. As far as I can see it is only used in tests.

**Review by DaedalusG — APPROVED** [review]

**Pull request merged**